src/Corax/Pipeline/Parsing/ScalarTransformers.cs

```diff
@@ -0,0 +1,191 @@
+
+                // dest[pos+1] = source[pos+1] | lowercaseTable[source[pos+1]]
+                ref byte d2 = ref Unsafe.AddByteOffset(ref destStart, pos + 1);
+                byte b2 = Unsafe.AddByteOffset(ref sourceStart, pos + 1);
+                d2 = (byte)(b2 | Unsafe.AddByteOffset(ref lowerCaseTable, b2));
+
+                // dest[pos+2] = source[pos+2] | lowercaseTable[source[pos+2]]
+                ref byte d3 = ref Unsafe.AddByteOffset(ref destStart, pos + 2);
+                byte b3 = Unsafe.AddByteOffset(ref sourceStart, pos + 2);
+                d3 = (byte)(b3 | Unsafe.AddByteOffset(ref lowerCaseTable, b3));
+
+                // dest[pos+3] = source[pos+3] | lowercaseTable[source[pos+3]]
+                ref byte d4 = ref Unsafe.AddByteOffset(ref destStart, pos + 3);
+                byte b4 = Unsafe.AddByteOffset(ref sourceStart, pos + 3);
+                d4 = (byte)(b4 | Unsafe.AddByteOffset(ref lowerCaseTable, b4));
+            }
+
+            switch (len - pos)
+            {
+                case 3:
+                    ref byte dp2 = ref Unsafe.AddByteOffset(ref destStart, pos + 2);
+                    byte bp2 = Unsafe.AddByteOffset(ref sourceStart, pos + 2);
+                    dp2 = (byte)(bp2 | Unsafe.AddByteOffset(ref lowerCaseTable, bp2));
+                    goto case 2;
+                case 2:
+                    ref byte dp1 = ref Unsafe.AddByteOffset(ref destStart, pos + 1);
+                    byte bp1 = Unsafe.AddByteOffset(ref sourceStart, pos + 1);
+                    dp1 = (byte)(bp1 | Unsafe.AddByteOffset(ref lowerCaseTable, bp1));
+                    goto case 1;
+                case 1:
+                    ref byte dp0 = ref Unsafe.AddByteOffset(ref destStart, pos);
+                    byte bp0 = Unsafe.AddByteOffset(ref sourceStart, pos);
+                    dp0 = (byte)(bp0 | Unsafe.AddByteOffset(ref lowerCaseTable, bp0));
+                    goto case 0;
+                case 0:
+                    break;
+            }
+
+            if (tokens != destTokens)
+                tokens.CopyTo(destTokens);
+
+            // We need to shrink the tokens and bytes output. 
+            destTokens = destTokens.Slice(0, tokens.Length);
+            dest = dest.Slice(0, source.Length);
+
+            return source.Length;
+        }
+
+        public static int ToLowercase(ReadOnlySpan<byte> source, ReadOnlySpan<Token> tokens, ref Span<byte> dest, ref Span<Token> destTokens)
+        {
+            nint sourcePos = 0;
+            nint destPos = 0;
+            nint len = source.Length;
+
+            ref byte bufferStart = ref MemoryMarshal.GetReference(source);
+            ref byte destStart = ref MemoryMarshal.GetReference(dest);
+            ref byte lowerCaseTable = ref MemoryMarshal.GetReference(StandardTransformers.LowercaseTable);
+
+            const int N64 = sizeof(ulong);
+            
+            while (sourcePos < len)
```

Comment:
Do we need the `while` here? After the first non-ASCII occurrence or when the inner `while` condition is `false`, we're sending all remaining bytes into `Rune` anyway.

Reply:
If I am reading it properly we are only sending non-ASCII characters to be processed by Runes system. Any bunch of ASCII characters are going to be processed using the traditional loop. Am I missing something here?